opm/autodiff/BlackoilModelBase_impl.hpp

```diff
@@ -1939,16 +1939,28 @@ namespace detail {
             if (std::isnan(mass_balance_residual[idx])
                 || std::isnan(CNV[idx])
                 || (idx < np && std::isnan(well_flux_residual[idx]))) {
-                OPM_THROW(Opm::NumericalProblem, "NaN residual for phase " << materialName(idx));
+                const auto msg = std::string("NaN residual for phase ") +  materialName(idx);
+                if (terminal_output_) {
+                    OpmLog::problem(msg);
+                }
+                OPM_THROW_NOLOG(Opm::NumericalProblem, msg);
```

Comment:
These are a lot of lines. I am wondering whether yet another Macro would make sense that only logs if a condition is true.
```
OPM_THROW_AND_LOG_IF(exception, message, condition);
```